src/rules/valid-expect.js

```diff
@@ -5,23 +5,118 @@
  * MIT license, Tom Vincent.
  */
 
-const { getDocsUrl } = require('./util');
+const {
+  getDocsUrl,
+  expectCase,
+  expectRejectsCase,
+  expectResolvesCase,
+  expectNotRejectsCase,
+  expectNotResolvesCase,
+} = require('./util');
 
 const expectProperties = ['not', 'resolves', 'rejects'];
+const promiseArgumentTypes = ['CallExpression', 'ArrayExpression'];
+
+/**
+ * expect statement can have chained matchers.
+ * We are looking for the closest CallExpression
+ * to find `expect().x.y.z.t()` usage.
+ *
+ * @Returns CallExpressionNode
+ */
+const getClosestParentCallExpressionNode = node => {
+  if (!node || !node.parent || !node.parent.parent) {
+    return null;
+  }
+
+  if (node.parent.type === 'CallExpression') {
+    return node.parent;
+  }
+  return getClosestParentCallExpressionNode(node.parent);
+};
+
+/**
+ * Async assertions might be called in Promise
+ * methods like `Promise.x(expect1)` or `Promise.x([expect1, expect2])`.
+ * If that's the case, Promise node have to be awaited or returned.
+ *
+ * @Returns CallExpressionNode
+ */
+const getPromiseCallExpressionNode = node => {
+  if (
+    node &&
+    node.type === 'ArrayExpression' &&
+    node.parent &&
+    node.parent.type === 'CallExpression'
+  ) {
+    node = node.parent;
+  }
+
+  if (
+    node.type === 'CallExpression' &&
+    node.callee &&
+    node.callee.type === 'MemberExpression' &&
+    node.callee.object.name === 'Promise' &&
+    node.parent
+  ) {
+    return node;
+  }
+
+  return null;
+};
+
+const checkIfValidReturn = (parentCallExpressionNode, allowReturn) => {
+  const validParentNodeTypes = ['ArrowFunctionExpression', 'AwaitExpression'];
+  if (allowReturn) {
+    validParentNodeTypes.push('ReturnStatement');
+  }
+
+  return validParentNodeTypes.includes(parentCallExpressionNode.type);
+};
+
+const promiseArrayExceptionKey = ({ start, end }) =>
+  `${start.line}:${start.column}-${end.line}:${end.column}`;
 
 module.exports = {
   meta: {
     docs: {
       url: getDocsUrl(__filename),
     },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          alwaysAwait: {
+            type: 'boolean',
+            default: false,
+          },
+          allowPromiseMethods: {
+            type: 'boolean',
+            default: false,
```

Comment:
should this default to true? I don't care about `Promise.resovle`, but doing `return expect(Promise.all([prom1, prom2])).resolves.toEqual([val1, val2])` seems perfectly valid to me

Reply:
But that's a different case. It shouldn't trigger any warnings on that anyways (if alwaysAwait: false). 

allowPromiseMethods are just for checking if `expect().resolves` was used in Promise.all, not the other way around.

Reply:
I think you got confused by the same thing as me - with `false`, this will already be allowed. With `true`, *any* `Promise.all`s will be allowed, even if they are not awaited or returned. TBH I have no idea why anyone would set it to true and not just eslint-disable if there's ever a use case for that.
Seeing that it was not just confusing for me and that IMO it makes the examples look like documentation mistakes, maybe we should just remove the setting?

Reply:
```js
Promise.all([
  expect(prom1).resolves.toEqual(val1),
  expect(prom2).resolves.toEqual(val2),
]);
```

should also always be valid

Reply:
@jeysal if somebody would have test cases using `done()` like below then they may wanna disable that check.

```js
Promise.all([
  expect(prom1).resolves.toEqual(val1),
  expect(prom2).resolves.toEqual(val2),
]).then(() => {
// some other operations & assertions
done();
});
```

@SimenB that's also valid. There shouldn't be any warnings with the examples you shared.

Reply:
> With `true`, _any_ `Promise.all`s will be allowed, even if they are not awaited or returned. 

Wat, them seems really weird. Adding `Promise.all` (or `Promise.resolve`, albeit that's weird) around the `expect` should make absolutely zero difference to this rule - it is about correctness, not style. The reason I brought is up is that I want us to understand it's the wrapping promise we want to be `return`ed or `await`ed, not for us to have any opinion on whether it's used or not

Reply:
Ok, read the docs, and tests, more thoroughly, an it does indeed behave the way I want it to. 😀 

However, it only does so provided `allowPromiseMethods` is not set to `false`. I think we should kill that option, I see no use cases for it. Can you describe one?

Reply:
> The reason I brought is up is that I want us to understand it's the wrapping promise we want to be returned or awaited, not for us to have any opinion on whether it's used or not

@SimenB I agree. However since promises can be fulfilled with `.then` statements as well and developers can use `done()` to conclude their tests. That's why somebody can want to have the rule in general but just want to disable it for async assertions that were wrapped with Promise methods. See my comments above as an example :)

Reply:
I think an `eslint-disable-next-line jest/valid-expect` is better in that case.

Reply:
@jeysal @SimenB I removed the option. I agree that in most cases nobody will need it. 

If that case annoys lots of people solution is already here. Somebody can revert the last commit in future if we need that option :D